jaxlib/pytree.cc

```diff
@@ -0,0 +1,616 @@
+    py::tuple tuple = py::reinterpret_borrow<py::tuple>(handle);
+    node.kind = Kind::kTuple;
+    node.arity = tuple.size();
+    for (py::handle entry : tuple) {
+      FlattenHelper(entry, leaves, tree);
+    }
+  } else if (PyList_CheckExact(handle.ptr())) {
+    py::list list = py::reinterpret_borrow<py::list>(handle);
+    node.kind = Kind::kList;
+    node.arity = list.size();
+    for (py::handle entry : list) {
+      FlattenHelper(entry, leaves, tree);
+    }
+  } else if (PyDict_CheckExact(handle.ptr())) {
+    py::dict dict = py::reinterpret_borrow<py::dict>(handle);
+    py::list keys = py::reinterpret_steal<py::list>(PyDict_Keys(dict.ptr()));
+    if (PyList_Sort(keys.ptr())) {
+      throw std::runtime_error("Dictionary key sort failed.");
+    }
+    for (py::handle key : keys) {
+      FlattenHelper(dict[key], leaves, tree);
+    }
+    node.kind = Kind::kDict;
+    node.arity = dict.size();
+    node.node_data = std::move(keys);
+  } else if ((node.custom = CustomNodeRegistry::Lookup(handle.get_type()))) {
+    node.kind = Kind::kCustom;
+    py::tuple out = py::cast<py::tuple>(node.custom->to_iterable(handle));
+    if (out.size() != 2) {
+      throw std::runtime_error(
+          "PyTree custom to_iterable function should return a pair");
+    }
+    node.node_data = out[1];
+    node.arity = 0;
+    for (py::handle entry : py::cast<py::iterable>(out[0])) {
+      ++node.arity;
+      FlattenHelper(entry, leaves, tree);
+    }
+  } else if (py::isinstance<py::tuple>(handle) &&
+             py::hasattr(handle, "_fields")) {
+    // We can only identify namedtuples heuristically, here by the presence of
+    // a _fields attribute.
+    py::tuple tuple = py::reinterpret_borrow<py::tuple>(handle);
+    node.kind = Kind::kNamedTuple;
+    node.arity = tuple.size();
+    node.node_data = py::reinterpret_borrow<py::object>(tuple.get_type());
+    for (py::handle entry : tuple) {
+      FlattenHelper(entry, leaves, tree);
+    }
+  } else {
+    node.kind = Kind::kLeaf;
+    leaves->append(py::reinterpret_borrow<py::object>(handle));
+  }
+  node.num_nodes = tree->traversal_.size() - start_num_nodes + 1;
+  node.num_leaves = leaves->size() - start_num_leaves;
+  tree->traversal_.push_back(std::move(node));
+}
+
+/*static*/ std::pair<py::list, std::unique_ptr<PyTreeDef>> PyTreeDef::Flatten(
+    py::handle x) {
+  py::list leaves;
+  auto tree = absl::make_unique<PyTreeDef>();
+  FlattenHelper(x, &leaves, tree.get());
+  return std::make_pair(std::move(leaves), std::move(tree));
+}
+
+py::object PyTreeDef::Unflatten(py::iterable leaves) const {
+  std::vector<py::object> agenda;
+  auto it = leaves.begin();
+  for (const Node& node : traversal_) {
+    if (agenda.size() < node.arity) {
+      throw std::logic_error("Too few elements for TreeDef node.");
+    }
+    switch (node.kind) {
+      case Kind::kLeaf:
+        if (it == leaves.end()) {
+          throw std::invalid_argument("Too few leaves for PyTreeDef");
```

Comment:
Include the size of `leaves` and the expected number of leaves (`traversal_.end().num_leaves()` or something)